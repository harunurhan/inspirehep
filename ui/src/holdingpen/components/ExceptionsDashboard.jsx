import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'antd';
import ExceptionsTable from '../components/ExceptionsTable';
import InlineList from '../../common/components/InlineList';
import './ExceptionsDashboard.scss';

class ExceptionsDashboard extends Component {
  static getDerivedStateFromProps(nextProps, prevState) {
    const { exceptions } = nextProps;
    const countEntriesByCollection = ExceptionsDashboard.getExceptionCountEntriesByCollection(
      exceptions
    );
    return {
      ...prevState,
      countEntriesByCollection,
    };
  }

  static getExceptionCountEntriesByCollection(exceptions) {
    const countMap = exceptions.reduce((counts, exception) => {
      if (counts[exception.collection] === undefined) {
        counts[exception.collection] = 0;
      }
      counts[exception.collection] += 1;
      return counts;
    }, {});
    return Object.entries(countMap);
  }

  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div>
        <h1>Exceptions Dashboard</h1>

        <div className="collections">
          <InlineList
            items={this.state.countEntriesByCollection}
            renderItem={([collectionName, collectionCount]) => (
              <span>
                {collectionName} {collectionCount}
              </span>
            )}
          />
        </div>

        <Row type="flex" justify="space-around">
          <Col>
            <ExceptionsTable exceptions={this.props.exceptions} />
          </Col>
        </Row>
      </div>
    );
  }
}

ExceptionsDashboard.propTypes = {
  exceptions: PropTypes.arrayOf(
    PropTypes.shape({
      collection: PropTypes.string,
      error: PropTypes.string,
      recid: PropTypes.number,
    })
  ).isRequired,
};

export default ExceptionsDashboard;
