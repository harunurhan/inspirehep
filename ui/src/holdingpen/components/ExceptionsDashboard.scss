.__ExceptionsDashboard__ {
  .collection-counts {
    text-align: center;
    margin-bottom: 12px;
    .space-around {
      margin: 0 24px;
    }
  }
}
