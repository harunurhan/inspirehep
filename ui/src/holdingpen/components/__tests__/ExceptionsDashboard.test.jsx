import React from 'react';
import { shallow } from 'enzyme';
import ExceptionsDashboard from '../ExceptionsDashboard';

const exceptions = [
  {
    collection: 'Job',
    error:
      "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
    recid: 1512550,
  },
  {
    collection: 'Hep',
    error:
      "Failed validating u'format' in schema[u'properties'][u'authors'][u'items'][u'properties'][u'emails'][u'items']:",
    recid: 1238165,
  },
  {
    collection: 'Conferences',
    error:
      "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
    recid: 1356791,
  },
  {
    collection: 'Hep',
    error:
      "Failed validating u'pattern' in schema[u'properties'][u'persistent_identifiers'][u'items'][u'properties'][u'value']:",
    recid: 1635467,
  },
];

describe('ExceptionsDashboard', () => {
  it('renders with all props set', () => {
    const wrapper = shallow(<ExceptionsDashboard exceptions={exceptions} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should calculate number of errors in each collection type', () => {
    const actualResult = ExceptionsDashboard.getExceptionCountEntriesByCollection(
      exceptions
    );
    const expectedResult = [['Job', 1], ['Hep', 2], ['Conferences', 1]];
    expect(actualResult).toEqual(expectedResult);
  });
});
