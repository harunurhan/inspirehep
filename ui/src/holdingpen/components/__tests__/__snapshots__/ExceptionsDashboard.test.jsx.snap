// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExceptionsDashboard renders with all props set 1`] = `
<div>
  <h1>
    Exceptions Dashboard
  </h1>
  <div
    className="collections"
  >
    <InlineList
      extractKey={[Function]}
      items={
        Array [
          Array [
            "Job",
            1,
          ],
          Array [
            "Hep",
            2,
          ],
          Array [
            "Conferences",
            1,
          ],
        ]
      }
      label={null}
      renderItem={[Function]}
      suffix={null}
    />
  </div>
  <Row
    gutter={0}
    justify="space-around"
    type="flex"
  >
    <Col>
      <ExceptionsTable
        exceptions={
          Array [
            Object {
              "collection": "Job",
              "error": "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
              "recid": 1512550,
            },
            Object {
              "collection": "Hep",
              "error": "Failed validating u'format' in schema[u'properties'][u'authors'][u'items'][u'properties'][u'emails'][u'items']:",
              "recid": 1238165,
            },
            Object {
              "collection": "Conferences",
              "error": "Failed validating u'format' in schema[u'properties'][u'urls'][u'items'][u'properties'][u'value']:",
              "recid": 1356791,
            },
            Object {
              "collection": "Hep",
              "error": "Failed validating u'pattern' in schema[u'properties'][u'persistent_identifiers'][u'items'][u'properties'][u'value']:",
              "recid": 1635467,
            },
          ]
        }
      />
    </Col>
  </Row>
</div>
`;
